File: leetcode/leetcode_cpp/average-of-levels-in-binary-tree.cpp
/*
https://leetcode.com/problems/average-of-levels-in-binary-tree

Solution1. DFS with list
time: o(n)
space: o(n)

Solution2. BFS with queue or stack
time: o(n)
space: o(n)

*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<pair<double,int>> l;
    void dfs(TreeNode* node, int level) {
        if (!node) return;

        if (l.size() <= level) l.push_back({node->val, 1});
        else {
            l[level].first += node->val;
            l[level].second += 1;
        }

        dfs(node->left, level+1);
        dfs(node->right, level+1);
    }

    vector<double> averageOfLevels_dfs(TreeNode* root) {
        if (!root) return {};

        dfs(root, 0);
        vector<double> res(l.size());
        for (int i=0; i<l.size(); ++i) {
            res[i] = (double)l[i].first / l[i].second;
        }

        return res;
    }

    vector<double> averageOfLevels_bfs(TreeNode* root) {
        if (!root) return {};

        queue<TreeNode*> q;
        q.push(root);

        vector<double> res;
        while (q.size()) {
            int qSize = q.size();
            double sum = 0;
            for (int i=0; i<qSize; ++i) {
                auto node = q.front();
                q.pop();
                sum += node->val;

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            res.push_back(sum/qSize);
        }

        return res;
    }

    vector<double> averageOfLevels(TreeNode* root) {
        return averageOfLevels_bfs(root);
    }
};

File: leetcode/leetcode_python/average-of-levels-in-binary-tree.py
'''
https://leetcode.com/problems/average-of-levels-in-binary-tree
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels_dfs(self, root: Optional[TreeNode]) -> List[float]:
        sums = []        
        def dfs(node, level):
            if not node:
                return
            if len(sums) <= level:
                sums.append([node.val, 1])
            else:
                sums[level][0] += node.val
                sums[level][1] += 1
            dfs(node.left, level+1)
            dfs(node.right, level+1)

        dfs(root, 0)
        res = []
        for v in sums:
            res.append(v[0] / v[1])

        return res

    def averageOfLevels_bfs(self, root: Optional[TreeNode]) -> List[float]:
        q = deque()
        q.append(root)

        res = []
        while q:
            sum = 0
            qSize = len(q)
            for _ in range(qSize):
                node = q.popleft()
                sum += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            res.append(sum/qSize)
        return res

    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        return self.averageOfLevels_bfs(root)